[STUD-490] add cluster api support

rmyers opened the pull request:
Adding support for stardog cluster api calls:

GET /admin/cluster - cluster info
GET /admin/cluster/status - cluster status

I have added some tests that mock the cluster responses as it seems that setting up a cluster for testing will require a lot of setup. I think this can be done with a new docker image or setup the existing one to create a cluster.

I based this off of the 'fix tests for 7' branch so probably easiest to look at the last commit for reviewing this.

jmrog commented:
@rmyers For the purposes of cleaner review, can you rebase commit `a609373` on the `v2` branch when you have a chance? The "fix tests" branch has been merged into `v2` now. Thanks!

jmrog commented on test/clusterInfo.spec.js:
```diff
@@ -0,0 +1,55 @@
+      }
+    );
+
+    done();
```
We try to avoid having to call `done` manually in the stardog.js tests (mostly just another convention, but it does help a little bit to avoid the "oops, I forgot" case, and it also removes the need to write your own `catch` clause). With promises, Jest can handle this itself. You could do:

```javascript
const clusterPromise = cluster.info(conn);
expect(Fetch.fetch).toHaveBeenCalledTimes(1);
/* . . . etc . . . */
return clusterPromise.then(res => { /* . . . etc . . . */ });
```

jmrog commented on test/optimizeDB.spec.js:
```diff
@@ -24,7 +24,10 @@ describe('optimizeDB()', () => {
     }));
 
-  it('should optimize an online DB', () =>
+  // TODO: This case is currently skipped because there is no way in Stardog
+  // 7.3.x to ensure that the DB is ready for optimization (so, this often
+  // errors out). The erroring out should be fixed in Stardog 7.4.1.
+  it.skip('should optimize an online DB', () =>
```
This part will need to be removed from the PR, as it's stale.

jmrog reviewed (changes requested):
Looks good, but a few requests for changes. Most of them are inline, but here's an additional one: you'll need to add type information for the `cluster` calls to "index.d.ts".

rmyers commented on lib/cluster.js:
```diff
@@ -0,0 +1,24 @@
+module.exports = {
+  info,
+  status,
+};
```
I changed the import to make it possible to mock the fetch function. Since it needs an Object to spyOn, couldn't get it to work with a direct import. But I'm working on getting the cluster to run with docker compose and then I can change the tests to a full integration instead of mocks to address this.

rmyers replied:
I was able to get a simple cluster to run and have updated the tests to use the cluster nodes for the connection without mocking fetch. This should address the code style issues.

rmyers commented:
@jmrog I have added a sample docker-compose and addressed the pr comments. When you have time please have a look. I'm not exactly sure how to get this going in circleci as I'm guessing that would break all the pr's and branches that don't have this code. You can however run this locally to test it out:

```bash
docker-compose -f .circleci/docker-compose.yml up 
```

then in another terminal run:

```bash
npm run test:cluster
```

jmrog commented on test/cluster/clusterInfo.spec.js:
```diff
@@ -0,0 +1,34 @@
+
+    return cluster.info(coordinatorConnection).then(res => {
+      const { nodes } = res.body;
+      const coordinator = nodes.find(node => node.search(/5821/) > 0);
+      const participant = nodes.find(node => node.search(/5822/) > 0);
```
Nitpick-ish: Instead of `node.search(regex) > 0`, you can just more directly do `node.includes(number)` here to accomplish the same purpose a little more clearly.

jmrog commented on test/cluster/clusterStatus.spec.js:
```diff
@@ -0,0 +1,42 @@
+
+    return cluster.status(coordinatorConnection).then(res => {
+      const { nodes } = res.body;
+      const coordinator = nodes.find(node => node.address.search(/5821/) > 0);
+      const participant = nodes.find(node => node.address.search(/5822/) > 0);
```
Same comment here (and below) re: `includes`.

jmrog reviewed (changes requested):
This looks great! I made a nitpicky comment inline that I'd like to have addressed before merging, but otherwise seems ready to go. (I can handle updating CircleCI-related stuff as necessary.)

jmrog reviewed (approved):
LGTM
